feature: useWatch/useFormState exact prop

bluebill1049 opened the pull request:
close #6977

```tsx
useWatch({
 name: 'test',
 exact: true
})

useFormState({
 name: 'test',
 exact: true
})
```

bluebill1049 committed:
feature: useWatch/useFormState exact prop

bluebill1049 committed:
update packages

bluebill1049 committed:
Merge branch 'master' into feature/exact
asdasdahhhhhkl

bluebill1049 committed:
packages update

bluebill1049 committed:
include unit tests coverage

bluebill1049 committed:
upgrade to react router 6

bluebill1049 committed:
Merge branch 'master' into feature/exact

bluebill1049 committed:
Merge branch 'master' into feature/exact

bluebill1049 committed:
fix lint error

barrymay commented on src/__tests__/useFormState.test.tsx:
```diff
@@ -599,4 +599,30 @@ describe('useFormState', () => {
+
+    fireEvent.blur(screen.getByRole('textbox'));
+
+    expect(screen.queryByText('touched')).toBeNull();
```
can you add something so touched becomes visible in the test?

bluebill1049 replied:
will do 👍

barrymay commented on src/logic/isWatched.ts:
```diff
@@ -0,0 +1,11 @@
+  !isBlurEvent &&
+  (_names.watchAll ||
+    _names.watch.has(name) ||
+    _names.watch.has((name.match(/\w+/) || [])[0]));
```
How is this line tested?

bluebill1049 replied:
mostly through integration tests, but yea worth adding a unit test on this one.

bluebill1049 replied:
good one, actually found a bug with it, will use `startsWith` instead.

barrymay replied:
If you want more complex, I'm happy to help build the Regex, I have a lot of practice there 😉

bluebill1049 replied:
take over, please! 👍 I will submit my breaking test.

barrymay replied:
Cool - I can finish it can you describe but exactly the problem is you're fixing? If you have any other tests that cover the situation, even if they break, that's helpful.  I'll sort it out tomorrow

bluebill1049 committed:
Merge branch 'master' into feature/exact

bluebill1049 committed:
test coverage for toggle exact

bluebill1049 committed:
breaking test

bluebill1049 committed:
update on the testing

barrymay committed:
Improve tests around isWatched (#7041)

bluebill1049 committed:
logic compress

bluebill1049 committed:
Merge branch 'master' into feature/exact

barrymay commented on src/logic/isWatched.ts:
```diff
@@ -0,0 +1,13 @@
+  (_names.watchAll ||
+    _names.watch.has(name) ||
+    [..._names.watch].some((watchName) =>
+      /^\.\w+/.test(name.slice(watchName.length)),
```
The reduction is good but we're missing the match of the starting text which is important.  I found a way to bring the existing text into the regex so it matches my meaning so I'll do that, but I can't sign this off. (I'll add the change in a bit)

Alternatively we can add the startsWith back but both likely add the same number of bytes back in.  (Nothing egregious but I think managing having concise code here is a balancing act - fully support the ... and .some move tho!)

barrymay replied:
PR to fix is in 😊

barrymay committed:
Feature/exact PR startsWith fix (#7042)

* Improve tests around isWatched

* Merge branch 'feature/exact' of git://github.com/react-hook-form/react-hook-form into feature/exact
- Adds back StartsWith check (fast for perf) and additional test to watch out for mismatched strings at the start

bluebill1049 commented:
all review comments are addressed 👍

bluebill1049 committed:
update packages

bluebill1049 committed:
update cypress 8

bluebill1049 committed:
combine logic to save bytes
